server/src/main/java/org/elasticsearch/cluster/routing/allocation/Hosts.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.routing.allocation;
+
+import org.elasticsearch.cluster.routing.RoutingNode;
+import org.elasticsearch.common.Strings;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * {@link Hosts} hosts records the mapping relationship between nodes and hosts,
+ * which makes it easy to find nodes on the same host
+ */
+public class Hosts {
+    private final Map<String, Host> node2Host;
```

Comment:
Really `Host` is just a placeholder for a network address which all the nodes on that host share - maybe we should just use the network address directly as a `String`?